File: src/main/java/net/bbelovic/adventofcode/day3/Day3PuzzlePart1.java
package net.bbelovic.adventofcode.day3;

import net.bbelovic.adventofcode.Puzzle;

import java.util.HashMap;
import java.util.Map;

class Day3PuzzlePart1 implements Puzzle<String, Integer> {

    private static final long VISITED_LOCATION = 1L;

    public Integer solve(String input) {
        var result = new HashMap<ArrayWrapper, Long>();
        int [] coordinates = {0, 0};
        result.put(new ArrayWrapper(coordinates), VISITED_LOCATION);
        var chars = input.toCharArray();
        for (char c: chars) {
            trackVisitedPositions(c, coordinates, result);
        }
        return result.size();
    }

    private void trackVisitedPositions(char c, int[] coordinates, Map<ArrayWrapper, Long> result) {
        if (c == '^') {
            coordinates[1] = coordinates[1] + 1;
        } else if (c == '>') {
            coordinates[0] = coordinates[0] + 1;
        } else if (c == 'v') {
            coordinates[1] = coordinates[1] - 1;
        } else if (c == '<') {
            coordinates[0] = coordinates[0] - 1;
        }
        ArrayWrapper actualPosition = new ArrayWrapper(coordinates);
        result.merge(actualPosition, VISITED_LOCATION, (oldValue, value)-> oldValue + VISITED_LOCATION);
    }
}

File: aoc-2018/src/main/kotlin/net/bbelovic/adventofcode/year2018/day3/Day3PuzzlePart1.kt
package net.bbelovic.adventofcode.year2018.day3

import net.bbelovic.adventofcode.Puzzle

class Day3PuzzlePart1 : Puzzle<List<String>, Long> {
    override fun solve(input: List<String>): Long {
        val ft = FabricTemplate()
        return input.asSequence()
                .map { it -> RectangleParser.parse(it) }
                .map { it -> ft.makeClaim(it) }
                .sum()
    }
}